x-pack/plugins/rule_registry/server/rule_data_client/index.ts

```diff
@@ -102,6 +103,51 @@ export class RuleDataClient implements IRuleDataClient {
     };
   }
 
+  async rolloverAliasIfNeeded({ namespace }: { namespace?: string }) {
+    const clusterClient = await this.getClusterClient();
+    const alias = getNamespacedAlias({ alias: this.options.alias, namespace });
+    const { body: simulatedRollover } = await clusterClient.indices.rollover({
+      alias,
+      dry_run: true,
+    });
+    const writeIndexMapping = await clusterClient.indices.get({
+      index: simulatedRollover.old_index,
+    });
+    const currentVersions = get(writeIndexMapping, [
+      'body',
+      simulatedRollover.old_index,
+      'mappings',
+      '_meta',
+      'versions',
+    ]);
+    const { body: simulatedIndex } = await clusterClient.indices.simulateIndexTemplate({
+      name: simulatedRollover.new_index,
+    });
+    const targetVersions = get(simulatedIndex, ['template', 'mappings', '_meta', 'versions']);
+    const componentTemplateRemoved =
+      Object.keys(currentVersions).length > Object.keys(targetVersions).length;
+    const componentTemplateUpdated = Object.entries(targetVersions).reduce(
+      (acc, [templateName, targetTemplateVersion]) => {
+        const currentTemplateVersion = get(currentVersions, templateName);
+        const templateUpdated =
+          currentTemplateVersion == null || currentTemplateVersion < Number(targetTemplateVersion);
+        return acc || templateUpdated;
+      },
+      false
+    );
+    const needRollover = componentTemplateRemoved || componentTemplateUpdated;
+    if (needRollover) {
+      await clusterClient.indices.rollover({
+        alias,
+        body: {
+          conditions: {
+            max_age: '5s',
+          },
+        },
+      });
+    }
```

Comment:
can we do this via a task? I don't really trust this (but I haven't thought about it long enough to be able to articulate my concerns better)

Reply:
We could look into alternatives to make this more robust. One option might be to specify the target index explicitly in the rollover request - I'll have to try it out and see if that makes the rollover request fail in the case where the target index already exists. If it does, then that should solve the race condition here since the rollover would only happen if another codepath hasn't rolled the index in between computing `needRollover` and actually rolling the index.

Reply:
There can be race conditions on a single kibana instance, but also by virtue of multiple kibana instances trying to roll over the index. It doesn't seem like a very unlikely scenario. A task will give us the guarantee that it will only happen once.

Reply:
Fixed in https://github.com/elastic/kibana/pull/102586/commits/19b9eaf3e0cdecc16de0c1bc76c6e8bd81103daa - by explicitly specifying the target index name pulled from the simulated rollover, we ensure that the rollover only happens if no other code path has rolled over the alias in the meantime.